src/sage/combinat/abstract_tree.py

```diff
@@ -1198,14 +1198,28 @@ def node_number(self):
             sage: T.node_number()
             10000
         """
-        count = 0
-
-        def incr(node):
-            nonlocal count
-            count += 1
-
-        self.iterative_pre_order_traversal(incr)
-        return Integer(count)
+        # An attribute _node_number for storing the number of nodes
+        # This is OK as AbstractTree is immutable
+        # If the attribute is not present, we compute it
+        try:
+            return self._node_number
+        except AttributeError:
+            pass
+        
+        if self.is_empty():
+            self._node_number = Integer(0)
+            return self._node_number
+        
+        def count(node):
+            if not node.is_empty():
+                # Using post-order
+                # Thus _node_number is computed for all non-empty subtrees
+                node._node_number = Integer(1)
+                node._node_number += sum([e._node_number for e in node
+                                             if not e.is_empty()])
```

Comment:
```suggestion
                node._node_number += sum(e._node_number for e in node
                                         if not e.is_empty())
```
We don't need the temporary list; the generator is sufficient.

Reply:
Yes, I think that we are good here.